### ISSUE TITLE
Use `native_toolchain_rust` to build ICU4X binary

### ISSUE THREAD

**Issue by mosuem** [dart-lang/i18n#857]
Instead of a manual [`Process.run('cargo'...`](https://github.com/dart-lang/i18n/blob/332b1a3b1f009e4315a7807301ea0537b559151a/pkgs/intl4x/hook/build.dart#L213), use [`package:native_toolchain_rust`](https://github.com/irondash/native_toolchain_rust/blob/main/example/dart_package/hook/build.dart).

fyi @knopp @dcharkes

**Comment by knopp** [comment]
This would certainly need some changes in `native_toolchain_rust`. I'll do my best to accommodate those. Missing pieces seem to be custom cargo arguments and static library. I'm not sure what the current the state of supporting static libraries in native assets? @dcharkes?

**Comment by mosuem** [comment]
Static linking will not be supported in the foreseeable future, but we are planning on producing a static library to remove unused symbols with the linker in the link hook, which produces the actual dylib being used at runtime.

**Comment by dcharkes** [comment]
> This would certainly need some changes in `native_toolchain_rust`. I'll do my best to accommodate those. Missing pieces seem to be custom cargo arguments and static library. I'm not sure what the current the state of supporting static libraries in native assets? @dcharkes?

Yes, the idea is to build static libs in the build hooks and then use the native linker to build a dynamic lib in the link hook. (At some point far in the future we might want to actually have the static lib linked to the AOT snapshot of Dart, but that requires a lot more changes.)

**Comment by mosuem** [comment]
@knopp - just curious, any changes in `native_toolchain_rust` so far to accomodate this use case? https://github.com/dart-lang/i18n/pull/866 adds even more rust stuff... :)

**Comment by knopp** [comment]
@mosuem, I published 0.1.0-dev.7 which supports extra cargo arguments. No support for static libraries, but I think I'd wait with that until there is actual support in native assets.

Is there something else required?

**Comment by mosuem** [comment]
Thanks - let me take a look.

> No support for static libraries, but I think I'd wait with that until there is actual support in native assets.

#866 will add static libraries. Dart doesn't support static linking, but we will build static libraries in the build hook, and then link them to get a dynamic library in the link hook. This is how we intend to treeshake symbols from the library.

**Comment by knopp** [comment]
> Thanks - let me take a look.
> 
> > No support for static libraries, but I think I'd wait with that until there is actual support in native assets.
> 
> #866 will add static libraries. Dart doesn't support static linking, but we will build static libraries in the build hook, and then link them to get a dynamic library in the link hook. This is how we intend to treeshake symbols from the library.

I don't think `native_toolchain_rust` can currently be used to do that. It always adds the asset (dynamic library) to the build output, what you're doing seems to be bit more exotic. Probably not a general use case.

**Comment by mosuem** [comment]
> It always adds the asset (dynamic library) to the build output

`native_toolchain_c` [solves that](https://github.com/dart-lang/native/blob/4eaec5a18977e215b3d37a6293564d1911137cef/pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart#L106) with an optional parameter in the `run` method. So I think it should be adaptable. But let me take a look, and I might file a PR. I agree that our use case is not super standard.

**Comment by knopp** [comment]
I see. That should be fairly easy to add. The remaining thing is the assumption that builder is always building a dylib.
